### PR TITLE
[READY] Space Dragon Rework

### PR DESCRIPTION
**Pull request by IndieanaJones**
## About The Pull Request

This PR reworks Space Dragon in a number of different ways, most noticeably from being an antag whose only objective was to kill a single head of staff to attempting to summon a massive horde of carp to the station to call the shuttle and end the round.  This and many more changes were made to make Space Dragon a much more enjoyable antag for both the dragon and crew, so let's get into it.

**WHAT WAS WRONG WITH SPACE DRAGON**
- Could tear through reinforced walls insanely fast, quickly leaving the station in shambles
- Fire breath went through glass and doors
- Only objective was to kill a single head of staff
- Very fast, could easily retreat into space, had no reason to stay aboard the station
- Commonly would tear apart the shuttle if it was alive when the shuttle was docked, and since it spawned late into the round and could idle in space, said situation was very likely
- Unless it was hostile, crew had no reason to kill it.  Space Dragon also had no reason to kill anyone but the head of staff selected
- Easily killed with any type of shotgun
- 50% resistance to lasers, the most common weapon that on-board security would have at any given time, making it extremely inefficient to use them on Space Dragon
- Tail whip wasn't very useful, only having a range of one and only being able to be used once every 60 seconds
- (more of a concern when we had cloning) Would gib killed opponents for health regen, making cloning impossible
- Enemies with fire resistance and good armor (miners) were practically immune to Space Dragon's attacks
- Generally seen as a worse but more common version of the slaughter demon event

**NEW BASE STATS AND ABILITIES**

- Max HP from 250 health to 400 health
- Added 30 armor piercing in melee
- Fire breath now deals 30 burn damage on contact (increased from 20)
- Fire breath deals 50 damage to mechs
- Fire breath no longer phases through windows and doors
- Resistance to lasers from 50% to 0%
- Now takes 4 seconds to tear through walls, 12 seconds to tear through reinforced walls (increased from instantly for both)
- No longer gibs dead bodies for health, instead swallows them whole like a snake.  Any bodies consumed will be regurgitated back out when Space Dragon dies
- Tail whip removed, replaced with...

**GUST ATTACK**

Space Dragon's new pushback ability.  Instead of using his flimsy tail to knock back enemies, Space Dragon rises into the air for a short bit, then uses his wings to create a force which stuns and knocks back friend and foe alike in a 3 tile radius.  The move doesn't come out immediately, but fast enough even with the animation tell to be successful.  Enemies hit are sent flying away from Space Dragon, and afterwards Space Dragon will be temporarily vulnerable and unable to act.  The first time this move is used, this downtime is relatively short, but the amount of time the tired animation lasts for is increased each time the move is used.  The move will return to it's original snappiness over time, but it must not be used in order for it to happen.  Currently, it takes about 30 seconds to undo the effects of one use of the move.  Because of this design, Space Dragon has no cooldown on this move, but the effects of spamming it will leave Space Dragon extremely vulnerable.

- All old objectives removed, adding instead...

**CARP RIFT**

Space Dragon has a new objective on spawning in, in which he attempts to create 3 rifts for a horde of space carp to travel through and swarm the station, destroying everything and reclaiming station space.  From spawning in, Space Dragon has 5 minutes to use his ability, Carp Rift, to open a rift on board the station.  Failure to do so within 5 minutes will result in Space Dragon returning from whence he came.  This rift can be placed on any location a cult summoning is allowed to occur, so Space Dragon can't cheese the crew by hiding it in the middle of open space.  Once spawned, a rift takes 4 minutes to fully charge.

The carp rift is a portal with 300 structure integrity.  It has resistances to bombing, and immunity from energy, biological, radiation, fire, and acid damage, with melee, lasers, and bullets dealing regular damage.  The portal cannot be moved once it is spawned in, but can be walked over.  Upon creation, the carp rift has 1 carp spawn, which ghosts can use to spawn in as Space Carp to assist Space Dragon with defending the rift.  Every 40 seconds, an extra carp spawn is added, and whenever the rift has a carp spawn available, it changes color from blue to purple.  At 2 minutes, a report will be made at Centcomm stating the location of the portal and the danger it poses, though ideally the crew should be looking for it once they hear the Space Dragon spawn announcement.  At 4 minutes, the portal reaches full charge, and become invulnerable to damage, along with becoming a goldish color.  Space Dragon is then given his ability to create another rift, and he must do so 3 times total.  After the 3rd time, all the charged rifts can spawn infinite sentient carp and the shuttle is auto-called akin to the Devil gamemode.  As of that point, Space Dragon has completed his objective.

**THE INTENDED GAMEPLAY LOOP**
- Space Dragon spawns, starts locating a place for the first rift
- Crew starts searching for rift
- Space Dragon spawns rift, begins defending it with space carp assistants.  Space Dragon makes choice of whether to go loud with placement or attempt to sneakily place to give the crew less time to attack it
- Rift is either destroyed or charged.  If charged, Space Dragon decides whether to stay in vicinity or try to hide the next rift

It should also be noted that if the currently charging rift is destroyed, Space Dragon isn't killed outright.  He instead has his speed set to 5 (very slow) and the amount of time he spends after a Gust Attack in stun is greatly increased, basically setting him up to die.  Any charged rifts are destroyed, and any minion carp alive will continue to persist as normal, though with 25 health, they will likely die soon after.

**OTHER CHANGES**

- Space Dragon's opening blurb has been rewritten, removing his ties to Wizards and moreso associating him with space carp.  While Space Dragon as a concept isn't entirely sci-fi, leaving it more ambiguous and tying him to space carp better fits the overall sci-fi theme.  Also, the blurb isn't a meme anymore either.

- Space Dragon is no longer a subtype of the ash drake, and therefore no longer a megafauna.  His information is stored inside his own file in the hostile folder.

- Space Dragon no longer shares his role preference with xenomorphs, instead having his own antag role and preference.

- Space Dragon's spawn weight has been increased from 8 to 10, and spawn time decreased from 70 minutes to 35 minutes (for reference, blob and nightmare have 10 spawn weight and spawn time set at 20 minutes)

## Why It's Good For The Game

Space Dragon was my first PR for this server and for SS13 in general.  Because of this, he ended up being a gimped version of the ash drake who would occasionally show up and either die or murder the station, with a bad objective and overall bad interaction with the station.  This PR intends to increase the quality of Space Dragon overall and make him a more fun and interesting antagonist for both sides.  It's also a personal ideal of mine that we might see of the midround antagonists changed to be made more fun for both the antags and crew, so I figured the best place to start is with the mistake I made myself.

## Changelog
:cl: Indie-ana Jones
tweak: Reports coming in indicate that the Space Dragon entity harassing stations has changed tactics, and is now attempting to flood stations with space carp.
tweak: Space Dragon preference is no longer shared with xenos, and is instead an independent option.
/:cl:

### CODE REVIEW

**Comment by Buggy123** [comment]
I like everything about this except the wall smash timers. Keep it as-is, wallsmashing is a good meme and it'll happen less anyway now that the dragon has a proper objective.

**Comment by Omni44** [comment]
As person that played quite a lot of Space Dragon (one of the more enjoyable antags I TCtrade most for) I like most of the changes, except 5 minute timer really, which I think puts a lot of pressure on what is otherwise freely roaming murder machine. I guess I understand nerf of wall smashing capabilities, especially since in general there's a push for more durable walls&windows, even though it was very fun thing to do. I'd rather let dragon still gib people as this makes it far more scary  and gruesome foe, but I'm probably alone here.

**Comment by Krysonism** [comment]
This seems nerf the dragon too much and the rift defense seems lame, especially since it ends the round.

I think space dragon is mostly fine as it is.

**Comment by IndieanaJones** [comment]
> This seems nerf the dragon too much and the rift defense seems lame, especially since it ends the round.

Would you be more open to it if charging the rift had some negative consequence for the rest of the round, instead of ending it outright?

**Comment by nianjiilical** [comment]
id rather the rift be less of a hard round ender and more of a passive round ender ala xenos

maybe have it level up the longer it's defended, starting with normal carp and eventually letting letting ghosts spawn as sharks/magicarps/etc

also seconding that 30s to kill an rwall is way too long

**Comment by IndieanaJones** [comment]
> id rather the rift be less of a hard round ender and more of a passive round ender ala xenos
> 
> maybe have it level up the longer it's defended, starting with normal carp and eventually letting letting ghosts spawn as sharks/magicarps/etc

Problem is that many people dislike the fact if xenos "win", they can delay rounds for long periods of time since nobody is available to call for the shuttle.

> also seconding that 30s to kill an rwall is way too long

That's valid, I'll wait for a few more opinions before I tweak anything though.

**Comment by HomocidlePancakes** [comment]
Dunno, it’s fun to be an incredibly powerful murder machine and attempt slaughter the station. This is kinda... Meh. Not a big fan of tower defense.

**Comment by IndieanaJones** [comment]
> Dunno, it’s fun to be an incredibly powerful murder machine and attempt slaughter the station. This is kinda... Meh. Not a big fan of tower defense.

We already have antags like this.  Wizard and Slaughter Demon are both examples of this, and now that cloning is out of the picture, having more sole murderbone antags seems like a waste.

**Comment by GuillaumePrata** [comment]
I had a quick look at the code and found no 'safety' for the spawned carps to survive random fire breaths or missclicks from the dragon at melee, not sure if it should be something they have to prevent team killing since they will try to defend a area together, just leaving it here.

**Comment by anconfuzedrock** [comment]
I prefer it trying to kill everyone but the rift "idea" doesn't seem so bad. Rift feels like it has double the health it should, but I don't see the dragon bein able to end the round as such a bad thing for how late in the round it comes. If you're not going to replace it with some carp portal storm or something...

The one change I think is needed is it's always annoying when half the issue with fighting some cult rift or whatever is that it's on the fucking ai sat and we don't have any all access memes going on. Restrict it to being closer to the center of the station more, or as a not boring idea:

Make it so the rift opens slower the farther it is from coords 127 127, since a central location would be better for flooding carp or whatever? So if it's in xenobiology on meta or something it'd take more like 15-20 minutes, while a ballsy carp placing it in the corporate showroom could get a mere 2 minute timer.

**Comment by Shadowflame909** [comment]
Feels like end game will tank it too much

Robotics will already have Mechs and everything, making repulse a joke.

Also. Bag of holding suicides.


Feels like another blob situation. Won't be very fun for the dragon.

**Comment by MonsWTF** [comment]
Considering there are already a handful of "guard one specific location" antags, what if the Space Dragon has to create and defend portals one at a time, until they have (3?) completed?

Reduce the total time on each, shift when the crew is notified about each portal, and give a few minutes after one portal is completed for the Space Dragon to establish the next one. Once the portal is complete, it's permanent and slowly spews out non-sentient carps.

If the Space Dragon does each in neighboring rooms, more exposure but more NPC defense; if they find another place to hide the second/third/_nth_ one, that just gives them a stealthy head-start.

**Comment by Iatots** [comment]
What if the dragon laid eggs and just wanted to turn the station into its lair? Perhaps the eggs could grow faster/bigger if near things made of gold or plasma.

**Comment by ATH1909** [comment]
Since you already have a mechanism in lace to keep space dragons from camping in space (they need to place a rift and defend it), can we return the default speed of space dragons to what it was before? If it can't move fast enough to close to melee range against people who try to kite it, the space dragon's melee attack becomes kind of useless and it just becomes a glorified fire breath turret, no?

**Comment by ATH1909** [comment]
I do like the idea of the space dragon being a soft round ender (with legions of infinitely respawning space carp) than a hard round ender (like a nuke op). Maybe force a shuttle call when the portal becomes immune to damage, so that the dragon and its carp army can have some time to have some fun ripping and tearing through the crew?

**Comment by IndieanaJones** [comment]
Alright, so after some analysis, here's my thoughts:
- Buff Space Dragon's HP, nerf rift health
- 3 rifts which take four minutes to charge instead of 1 for ten
- Revert speed change
- Buff wall-breaking, but still not make it instant
- Charged rifts intemitently spawn regular space carp
- Once all 3 rifts are charged, infinite sentient carp spawns
- (Maybe) buff damage against mechs

**Comment by anconfuzedrock** [comment]
more health AND speed for it might be a bad idea, but that's just what I think. Just make sure however it's tracking 'cant put the rift in the same place' isn't by room because that means it's going to put one in the ai core, one in ai sat tcomms, and one in the ai sat teleporter room. Don't think that fixes my concerns with it if that was the intent because it'll just pick three places that are annoying to get to so the crew has 2 minutes left by the time they're near it.

**Comment by Tharcoonvagh** [comment]
HELLO VORECODE?

**Comment by IndieanaJones** [comment]
> HELLO VORECODE?
The morph's been in the game for years and does the same exact thing.  Snakes eat their prey whole.  Space Dragon is snake-esque, thus it makes sense.

**Comment by ATH1909** [comment]
I feel like 4 sapient carp spawns per rift is a bit low, but I don't know what a better number would be. 6?

**Comment by gitbirb** [comment]
To encourage the space dragon going around and "hunting" like it should, perhaps you could add a benefit to killing and eating your foes? Such as healing the rift or slightly lowering the timer on the rift.

**Comment by IndieanaJones** [comment]
Alright, here's all the changes:
- Speed increased from 1 to 0
- Space Dragon must now summon 3 rifts in different locations to win.  The rifts take 4 minutes to charge each, and announce their location at 2 minutes.  One carp spawn is given at creation, and another is generated every 40 seconds
- Charged have a 2.5% chance to spawn a (non-sentient) space carp every second
- Changed the check for rift placement to use the cult's summoning check, since the blob is apparently allowed to activate on places such as the White Ship.
- Walls take 4 seconds to break, reinforced walls take 12 seconds to break
- Completing the objective will force a shuttle call instead of ending the round outright
- Nerfed rift integrity from 1000 to 300
- Fire breath does 50 damage to mechs
- Space Dragon's spawn weight increased from 8 to 10
- Space Dragon's minimum spawn time decreased from 70 minutes to 35 minutes

**Comment by IndieanaJones** [comment]
I think this swathe of changes should cover most of the critique given.  I have also made it so Space Dragon shows up more often and earlier now that he's more "proper", if anyone thinks I should set the rates differently, I'm open to suggestions.  However, as of the current time, I consider this PR ready.

File: code/controllers/subsystem/ticker.dm [hunk 1]
```diff
@@ -539,8 +539,7 @@ SUBSYSTEM_DEF(ticker)
 		if(CLOCK_PROSELYTIZATION)
 			news_message = "The burst of energy released near [station_name()] has been confirmed as merely a test of a new weapon. However, due to an unexpected mechanical error, their communications system has been knocked offline."
 		if(SHUTTLE_HIJACK)
-			news_message = "During routine evacuation procedures, the emergency shuttle of [station_name()] had its navigation protocols corrupted and went off course, but was recovered shortly after."
-
+			news_message = "During routine evacuation procedures, the emergency shuttle of [station_name()] had its navigation protocols corrupted and went off course, but was recovered shortly after."
```
Comment (spookydonut) [inline code/controllers/subsystem/ticker.dm]: whitespace changes?

Comment (IndieanaJones) [reply code/controllers/subsystem/ticker.dm]: Fixed

**Comment by IndieanaJones** [comment]
Added autodoc stuff to the space_dragon.dm file.

File: code/modules/mob/living/simple_animal/hostile/space_dragon.dm [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+	if(riftTimer == -1)
+		return
+	riftTimer = min(riftTimer + 1, maxRiftTimer + 1)
+	if(riftTimer == (maxRiftTimer - 60))
+		to_chat(src, "<span class='boldwarning'>You have a minute left to summon the rift!  Get to it!</span>")
+		return
+	if(riftTimer == maxRiftTimer)
+		to_chat(src, "<span class='boldwarning'>You've failed to summon the rift in a timely manner!  You're being pulled back from whence you came!</span>")
+		destroy_rifts()
+		QDEL_NULL(src)
```
[the remark this answers is not in the document]
Comment (IndieanaJones) [reply code/modules/mob/living/simple_animal/hostile/space_dragon.dm]: That was a mistake on my end, should just be a regular qdel.

**Comment by kriskog** [comment]
testmerged event hall and manuel

**Comment by kriskog** [comment]
```
[S] has opened a rift! (Orbit)
```

**Comment by kriskog** [comment]
Ends up in uncategorized in the traitor panel
image

**Comment by rdragan** [comment]
Played with this a bit during round 131950
Can't really talk balance/gameplay as I've never played with the old dragon. It was fun however.

There seemed to be a slight issue where the fire breath would play the sound but not actually shoot any fire. This happened about 1/10 fire breaths. Not sure if it has a cooldown, but even then, you might want it to not play the sound or play a different one for better feedback.

**Comment by Atexhoffman** [comment]
Played as a spess dragoon in round 132011, and melee attacks seem to do nothing to mechs of any kind.

**Comment by Daxxed** [comment]
Skog spawned one of these on Event Hall.
I got to play space carp for 5 seconds before being killed in one hit by an npc slime.
10/10 would plank again.

The idea is alot better than just a solo antag roaming the station, anything that brings more ghosts into the round gets an upvote from me

**Comment by IndieanaJones** [comment]
So, I'm going to assume the current balance is pretty good, since nobody's complained about it.  I've got some adjustments I'm going to push:
- Space Dragon's fire now goes through doors if the door is open (this is what rdragan was experiencing, since open firelocks would stop fire)
- Space Dragon's melee now does more damage against mechs
- Space Dragon opening rifts now displays properly
- Space Dragon appears in his own category in the antag panel.

**Comment by IndieanaJones** [comment]
Oh, and I guess we'll need to use the new spacewalk trait too.

**Comment by TheMidnightRose** [comment]
Kill healing feels weak, could we get a way to heal while sitting on a open rift? Similar to how blobbernauts heal on blob core/node.

**Comment by IndieanaJones** [comment]
Alright, after asking around Discord, turns out they believe the new dragon is actually pretty weak.  Now, since the dragon is a midround, its supposed to have some challenge but be able to be beaten by a good crew, so it not winning often is inline with balance.  However, apparently its very easy to beat it, so I'm going to buff it up a little bit with the following:
- Standing on top of the rift as Space Dragon or a carp passively heals you
- Upon fully charging a rift, Space Dragon and his carps will be teleported somewhere outside the station akin to SD's spawn points, and be healed.

**Comment by IndieanaJones** [comment]
So yeah, Space Dragon and carps can now heal if they share a tile with a carp, healing them 10 HP per second flat.  Space Dragon, after a rift fully charges, turns red, heals completely, and can move slightly faster for 30 seconds, allowing him to escape and find another rift location.

**Comment by IndieanaJones** [comment]
Ghosts now receive a notification when a carp spawn is available.

File: code/modules/events/space_dragon.dm [hunk 3]
```diff
@@ -2,8 +2,8 @@
 	name = "Spawn Space Dragon"
 	typepath = /datum/round_event/ghost_role/space_dragon
 	max_occurrences = 1
-	weight = 8
-	earliest_start = 70 MINUTES
+	weight = 10
```
Comment (kriskog) [inline code/modules/events/space_dragon.dm]: Too high, 5 maybe.

**Review by kriskog — CHANGES_REQUESTED** [review]

**Comment by adamkad1** [comment]
Charged rift doesnt seem to let you be sentient carp

**Comment by kriskog** [comment]
This is not the issue tracker, the PR is merged.

**Pull request merged**